src/ports.rs

```diff
@@ -0,0 +1,19 @@
+use std::net::{Ipv4Addr, SocketAddrV4, TcpListener};
+
+fn check_port_available(port: u16) -> bool {
+    let ipv4 = SocketAddrV4::new(Ipv4Addr::UNSPECIFIED, port);
+    TcpListener::bind(ipv4).is_ok()
+}
+
+pub fn validate_port_available(port: u16, server_name: &str) -> anyhow::Result<()> {
+    if check_port_available(port) {
```

Comment:
Possible TOCTOU problem here:

https://en.wikipedia.org/wiki/Time-of-check_to_time-of-use